File: leetcode/238_product_of_array_except_self.py
from typing import List


class Solution:

    def productExceptSelf(self, nums: List[int]) -> List[int]:
        N = len(nums)
        products = [1] * N
        curr = nums[0]
        for i in range(1, N):
            products[i] *= curr
            curr *= nums[i]
        curr = nums[-1]
        for i in reversed(range(0, N-1)):
            products[i] *= curr
            curr *= nums[i]
        return products

File: leetcode/java/238ProductOfArrayExceptSelf.java
/**
 * T: O(N) S: O(1)
 *
 * <p>To not use any intermediate space we remember the last seen multiplier and multiply every
 * element by it. Multiplication should be done before changing the multiplier to satisfy "except
 * self" requirement.
 */
import java.util.Arrays;

class Solution {
  public int[] productExceptSelf(int[] nums) {
    int[] multipliedNums = new int[nums.length];
    Arrays.fill(multipliedNums, 1);
    int multiplier = 1;
    for (int i = 0; i < nums.length; i++) {
      multipliedNums[i] *= multiplier;
      multiplier *= nums[i];
    }
    multiplier = 1;
    for (int i = nums.length - 1; i >= 0; i--) {
      multipliedNums[i] *= multiplier;
      multiplier *= nums[i];
    }
    return multipliedNums;
  }
}
